src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,651 @@
-            remap_trace: HashMap::new(),
-            since: since.clone(),
-            upper: Antichain::from_elem(Timestamp::minimum()),
-            source_upper: HashMap::new(),
-            write_handle,
-            read_handle,
-            listener,
-            now,
-            update_interval_ms: update_interval
-                .as_millis()
-                .try_into()
-                .expect("huge duration"),
+            upper: Antichain::from_elem(IntoTime::minimum()),
+            source_upper: MutableAntichain::new(),
+            remap_handle,
+            clock_stream,
         };
 
-        // Load the initial state that might exist in the shard
-        let initial_batch = operator.sync(&upper).await;
-
-        Ok((operator, initial_batch))
-    }
-
-    /// Compacts the internal state
-    pub async fn compact(&mut self, new_since: Antichain<Timestamp>) {
-        if !PartialOrder::less_equal(&self.since, &new_since) {
-            panic!(
-                "ReclockOperator: `new_since` ({:?}) is not beyond \
-                `self.since` ({:?}), using {:?} and {:?}",
-                new_since, self.since, self.read_handle, self.listener
-            );
-        }
-
-        for bindings in self.remap_trace.values_mut() {
-            // Compact the remap trace according to the computed frontier
-            for (timestamp, _) in bindings.iter_mut() {
-                timestamp.advance_by(new_since.borrow());
+        // Initialize or load the initial state that might exist in the shard
+        let trace_batch = if upper.elements() == [IntoTime::minimum()] {
+            let (_, upper) = operator.clock_stream.next().await.expect("end of time");
+            let batch = vec![(FromTime::minimum(), IntoTime::minimum(), 1)];
+            match operator.append_batch(batch, upper.clone()).await {
+                Ok(trace_batch) => trace_batch,
+                Err(Upper(actual_upper)) => operator.sync(actual_upper.borrow()).await,
             }
-            // And then consolidate
-            consolidation::consolidate(bindings);
-        }
-        self.since = new_since;
-        self.read_handle.maybe_downgrade_since(&self.since).await;
+        } else {
+            operator.sync(upper.borrow()).await
+        };
+
+        (operator, trace_batch)
     }
 
     /// Advances the upper of the reclock operator if appropriate
-    pub async fn advance(&mut self) -> ReclockBatch {
-        // Any updates to the remap trace that occured during advancing.
-        let mut trace_updates = vec![];
-
-        if self.next_mint_timestamp().is_ok() {
-            let empty: Vec<(PartitionId, MzOffset)> = Vec::new();
-            while let Err(Upper(actual_upper)) = self.append(&empty).await {
-                trace_updates.extend(self.sync(&actual_upper).await.updates);
+    pub async fn advance(&mut self) -> ReclockBatch<FromTime, IntoTime> {
+        match self.clock_stream.next().now_or_never() {
```

Comment:
hmmmmmmm `now_or_never` seems like a codesmell, it relies on the semantics of the underlying poll function...is there a way we can structure this differently? if its not easily possible then ignore me

Reply:
It relies on `.next()` being cancel safe, which it is, but I agree that it's not nice. There is a way to structure this differently but it's a big change so I think we should leave it as-is for now. I think what needs to happen is that instead of these operators being driven externally by calling various methods we should just make a timely operator that `tokio::selects!` its inputs and reacts accordingly to all the events.

Reply:
yeah lets just add a commend and leave it as is!